src/adapters/uniswap-v3/products/pool/uniswapV3PoolAdapter.ts

```diff
@@ -0,0 +1,348 @@
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'UniswapV3',
+      description: 'UniswapV3 defi adapter',
+      siteUrl: 'https://uniswap.org/',
+      iconUrl:
+        'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+    }
+  }
+
+  /**
+   * Update me.
+   * Add logic to build protocol token metadata
+   * For context see dashboard example ./dashboard.png
+   * We need protocol token names, decimals, and also linked underlying tokens
+   */
+  @CacheToFile({ fileKey: 'protocol-token' })
+  async buildMetadata() {
+    throw new Error('Implement me')
+
+    return {}
+  }
+
+  /**
+   * Update me.
+   * Returning an array of your protocol tokens.
+   */
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    throw new Error('Implement me')
+  }
+
+  /**
+   * Update me.
+   * Add logic to get userAddress positions in your protocol
+   */
+  async getPositions({
+    userAddress,
+    blockNumber,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const positionsManagerContract = PositionManager__factory.connect(
+      contractAddresses[this.chainId]!.positionManager,
+      this.provider,
+    )
+
+    const balanceOf = await positionsManagerContract.balanceOf(userAddress, {
+      blockTag: blockNumber,
+    })
+
+    const tokenIds = await Promise.all(
+      [...Array(Number(balanceOf)).keys()].map((index) =>
+        positionsManagerContract.tokenOfOwnerByIndex(userAddress, index, {
+          blockTag: blockNumber,
+        }),
+      ),
+    )
+
+    const positionMetadata = await Promise.all(
+      tokenIds.map((tokenId) =>
+        positionsManagerContract.positions(tokenId, { blockTag: blockNumber }),
+      ),
+    )
+
+    const nonZeroPositions = filterMap(positionMetadata, (position, index) => {
+      if (position.liquidity == 0n) return undefined
+
+      return {
+        liquidity: position.liquidity,
+        token0: position.token0,
+        token1: position.token1,
+        fee: position.fee,
+        tokenId: tokenIds[index],
+      }
+    })
+
+    if (nonZeroPositions.length == 0) return []
+
+    const tokenBalancesPromises = nonZeroPositions.map(
+      ({ liquidity, tokenId }) =>
+        positionsManagerContract.decreaseLiquidity.staticCallResult(
+          {
+            tokenId: tokenId as bigint,
+            liquidity,
+            amount0Min: 0n,
+            amount1Min: 0n,
+            deadline,
+          },
+          { from: userAddress, blockTag: blockNumber },
+        ) as unknown as Promise<
+          [bigint, bigint] & { amount0: bigint; amount1: bigint }
+        >, // IMO typechain has an bug here, incorrect type being returned
```

Comment:
typechain says the type being returned is a nested array but its not you can check in the function in the abi

Typechain says
[[bigint, bigint] & { amount0: bigint; amount1: bigint }] 

I say 
[bigint, bigint] & { amount0: bigint; amount1: bigint }

abi